### ISSUE TITLE
[WebGPU EP] Deterministic audio corruption in Kokoro-82M TTS on Intel Iris Xe — fp32 corrupts a subset of inputs, fp16 corrupts all; same model clean on WASM EP and on Qualcomm Adreno

### ISSUE THREAD

**Issue by maverick-marc** [microsoft/onnxruntime#29807]
Running the Kokoro-82M TTS model (`onnx-community/Kokoro-82M-v1.0-ONNX`) on the WebGPU EP
on **Intel Iris Xe** produces **deterministically corrupted audio** for certain inputs:

- **fp32**: roughly ⅓ of sentences render with a degraded "congested" timbre — measured as
  either elevated high-frequency noise (6kHz+ RMS share 0.19–0.42 vs ≤0.15 for clean renders
  of the same sentences) or mid-band (700–2000Hz) energy dropout.
- **fp16**: **every** input we tested renders corrupted (10/10 sentences, HF share 0.26–0.38
  vs 0.13–0.31 for paired WASM renders of the identical text).
- **q8 / q4f16** on WebGPU produce unintelligible output (phoneme garbage / noise) and q8 is
  also slower than the WASM EP — possibly a separate issue, included for completeness.

Key characteristics that should help triage:

1. **Deterministic and content-triggered**: the same input sentence produces byte-identical
   corrupt audio on every run (verified across independent sessions — identical per-sentence
   RMS statistics). Other sentences through the same session render perfectly. This is not a
   race/uninitialized-memory pattern; it looks like a numerical defect in specific kernel(s)
   exercised by specific input sequences.
2. **Hardware-specific**: the identical code + model + dtype on **Qualcomm Adreno 7xx**
   (Android, same Chromium 150) renders **completely clean** at fp32. The **WASM EP** on the
   affected machine is also completely clean.
3. **Timbre-level, not logic-level**: the corrupted audio remains 100% intelligible to
   whisper-large-v3 (perfect transcription) but is audibly degraded to humans — consistent
   with bounded numerical error rather than wrong control flow.

## To reproduce

```js
import { KokoroTTS } from "kokoro-js"; // 1.2.0 (bundles @huggingface/transformers 3.8.1 / ort-web 1.22.x)
const tts = await KokoroTTS.from_pretrained("onnx-community/Kokoro-82M-v1.0-ONNX", {
  device: "webgpu", dtype: "fp32",   // also try "fp16"
});
// Sentences that reproduce the fp32 corruption deterministically on Intel Iris Xe:
const bad = "Each sentence you hear was generated separately, so the timing log shows exactly how long every line took.";
const good = "The new engine runs on the graphics card instead.";
const audio = await tts.generate(bad, { voice: "af_bella" });
// Compare against device:"wasm" render of the same text: HF-noise share ~2x higher, audibly "congested".
```

**Attached: a clean/corrupt pair of the IDENTICAL sentence** ("Each sentence you hear was
generated separately, so the timing log shows exactly how long every line took."):
- `ort-issue-CORRUPT-intel-webgpu-fp32.mp4` — WebGPU EP fp32 render on Intel Iris Xe
- `ort-issue-CLEAN-wasm-same-sentence.mp4` — WASM EP render of the same text, same machine
The corrupt render reproduces byte-identically on every run. A minimal standalone repro page
(single HTML, kokoro-js from CDN + device/dtype picker + spectral scoring) available on request.

## Environment

- onnxruntime-web **1.22.x** (as bundled by @huggingface/transformers 3.8.1 / kokoro-js 1.2.0)
- **Affected:** Windows 11, Chromium 150 (Chrome and Brave), Intel **Iris Xe** (i7-1360P, Gen-12LP)
- **Not affected:** same code/model on Android Chromium 150, Qualcomm **Adreno 7xx** (clean at fp32);
  WASM EP clean everywhere (q8 and fp32)
- crossOriginIsolated=true (threaded WASM available); issue reproduces regardless of isolation

## Expected behavior

WebGPU EP output numerically consistent with WASM EP output for the same model/dtype
(within normal float tolerance — not audible timbre corruption).

## Additional context

We measured the corruption's spectral transfer (clean-vs-corrupt renders of identical text,
deterministic pairing): mid-band 300–2000Hz attenuated ~1.2–1.6dB, 4.5kHz+ boosted
+1.3–1.8dB with added noise (hiss variant), plus a second variant with mid-band dropout
(muffle). Happy to provide audio pairs, per-sentence spectral data, or run diagnostics
builds on the affected hardware.

https://github.com/user-attachments/assets/f3ccd31d-8c33-4073-b634-bc71da306ee6
https://github.com/user-attachments/assets/50faa1f0-3f21-44f4-ac96-4645d9d31c8f

**Comment by wuisabel-gif** [comment]
I went looking for a defect class that matches this fingerprint (fp16 corrupts all inputs, fp32 a subset, Intel Iris Xe only while Adreno/WASM are clean) and I think the standalone WebGPU **Softmax** kernel is a strong candidate.

In `onnxruntime/core/providers/webgpu/math/softmax.cc`, the online-softmax reduction keeps its running **max and sum in the tensor element type**, i.e. **fp16** when the model is fp16:

```wgsl
var thread_sum = x_element_t(0.0);                                  // fp16 accumulator
thread_sum = thread_sum * exp(thread_max - new_max) + x_element_t(SumVector(shifted_exp, components));
...
var<workgroup> thread_max_scalar_shared : array<x_element_t, wg>;   // fp16 shared reduction
var<workgroup> thread_sum_scalar_shared : array<x_element_t, wg>;
```

Summing a full row of `exp()` terms into an fp16 accumulator loses a lot of precision (fp16 has ~3 significant digits). Notably, the rest of the EP already avoids this:

- `contrib_ops/webgpu/bert/attention.cc` `InPlaceSoftmaxProgram` accumulates in **f32** (`thread_sum: array<f32>`, `thread_sum_local: f32`, values upcast to `f32_val_t`).
- `nn/layer_norm.cc` and `nn/instance_norm.cc` both accumulate their sums/variances in **f32**.

So the fp16 accumulation in the standalone Softmax op looks like an oversight rather than intentional.

Why it lines up with the report:
- **fp16 corrupts every input** – the softmax sum is always fp16-accumulated in fp16 mode.
- **fp32 only sometimes** – in fp32 `x_element_t` is f32, so this kernel is fine; the fp32 subset is likely a separate issue.
- **Intel Iris Xe only, Adreno/WASM clean** – this is the key part. Many GPUs (and the CPU/WASM path) evaluate fp16 arithmetic at higher internal precision, which hides the accumulation error; Intel Gen-12LP does honest fp16, so it surfaces as bounded, audible timbre corruption. That is consistent with "100% intelligible but degraded."

A Kokoro model exported via transformers.js decomposes attention into MatMul + Softmax + MatMul, so it exercises this exact Softmax op.

Proposed fix: accumulate the running max/sum in f32 in `SoftmaxProgram` and downcast only at the final write, mirroring `InPlaceSoftmaxProgram`. It is higher precision, consistent with the rest of the EP, and a small change.

I don't have Intel Iris Xe hardware to confirm this end-to-end. If someone who can reproduce is able to test a build with f32 softmax accumulation (or capture the pre-/post-softmax tensors from the affected sentences), that would confirm or rule it out. Happy to put up a PR with the change if this looks like the right direction.

**Comment by maverick-marc** [comment]
Ran a systematic sweep on the affected Intel Iris Xe (i7-1360P), rendering a short→long set of neutral sentences on fp32-WebGPU vs WASM on the same machine, scored by a worst-voiced-phrase 6 kHz high-pass RMS share. delta = gpuHF − wasmHF isolates the GPU-added corruption (WASM = clean baseline). vendor: intel, fp32 on WebGPU, byte-identical run-to-run.

3 of 12 corrupt (delta ≥ 0.05), all among the longest sentences; shorter ones clean:

words  gpuHF  wasmHF   delta   sentence
   4   0.130  0.127  +0.004   The kettle boiled quickly.
   8   0.210  0.222  -0.012   Rain fell on the quiet street after dusk.
   9   0.160  0.149  +0.011   She placed the heavy book back on the shelf.
   9   0.150  0.138  +0.012   A distant train rumbled slowly past the old station.
  10   0.140  0.136  +0.003   The lighthouse still guides ships along the rocky northern coast.
   9   0.179  0.168  +0.012   Autumn leaves drifted across the empty playground all afternoon.
  13   0.135  0.144  -0.009   She sells fresh shellfish and salted fish at the seaside stall each summer.
  18   0.206  0.179  +0.027   The museum's newest exhibit traces the history of written language ...   (mild)
  20   0.372  0.388  -0.016   After the storm finally passed, the villagers gathered in the square ...
  21   0.315  0.162  +0.152   Scientists carefully measured the temperature of the lake every hour ...   CORRUPT
  23   0.343  0.188  +0.155   The quick brown fox jumps over the lazy dog while a gentle breeze ...        CORRUPT
  23   0.403  0.166  +0.237   Every single morning the baker kneaded the soft dough by hand ...           CORRUPT

[ort29807-intel-fp32-pairs.zip](https://github.com/user-attachments/files/30361043/ort29807-intel-fp32-pairs.zip)

Length correlation is strong: Pearson r = 0.88 (gpuHF vs words), 0.73 (delta vs words); mean delta short (≤8w) ≈ 0 vs long (≥15w) = +0.111. All three corruptions are the longest sentences.

Two controls fell out of this and are worth noting: (1) a naturally sibilant 13-word sentence ("She sells fresh shellfish…") and a loud 20-word one both read high HF on both EPs with delta ≈ 0 / negative — i.e. not corrupted, confirming the effect is GPU-added HF, not just HF-heavy content; (2) the 18-word museum sentence shows only a mild sub-threshold elevation (+0.027), so onset isn't a hard length cutoff.

Net: on fp32 the corruption is content-gated with a strong length / reduction-width bias — consistent with @wuisabel-gif's read that the fp32 subset is separate from the fp16-softmax path while pointing at the same precision-in-reductions family (an accumulation cliff more exposed as reduction width grows).

Attached: ort29807-intel-fp32-pairs.zip — corrupt/clean WAV pair for each flagged sentence (identical text, *_gpu-fp32-CORRUPT.wav vs *_wasm-CLEAN.wav, 24 kHz) + the report.

We have the repro hardware and are glad to test a build with the f32 softmax accumulation (does it clean the fp32 subset too, or only fp16?), provide per-sentence spectral deltas, or run a diagnostics build.

**Comment by prathikr** [comment]
Hi @maverick-marc, online softmax is a recent optimization, you would not hit this in onnxruntime-web v1.22.x. Can you confirm if this issue is still present using onnxruntime 1.28? If yes, I can implement @wuisabel-gif's proposed fix.

**Comment by antoniosteiger** [comment]
I also observed this issue on multiple systems now, even the official kokoro webgpu demos don't work on an AMD RDNA3 7800 XT, nor intel integrated uhd graphics, nor an AMD laptop I had lying around. The audio produced with FP32 seems to resemble the sentence cadance, but sounds extremely muffled and is not understandable.
FP16 produces no audio on the 7800 XT. 
Unfortunately I don't have time to dig deeper, but I encourage the team to have a look at this again, since it seems more widespread than expected.
